tools/bench/parsec/evm/contracts/gen_header.py

```diff
@@ -19,104 +20,134 @@ def to_snake(name):
+
+        # Write the namespace for the contracts
+        f.write('namespace cbdc::parsec::evm_contracts {\n')
+
+        # The first 4 bytes of the input data sent to a contract is the
+        # method selector in ETH. It is the first 4 bytes of
+        # keccak256(<method signature>)
+        f.write('  static constexpr size_t selector_size = 4;\n')
+
+        # Parameters in a method call are always 32 bytes
+        f.write('  static constexpr size_t param_size = 32;\n\n')
+
+        # Since params are 32 bytes, addrs must be copied at a 12 bytes offset
+        f.write('  static constexpr size_t address_param_offset = 12; // in ABIs addresses are also 32 bytes\n')
+
+        # Generate methods for all contracts
+        for k, v in loaded_contracts.items():
+            # The data needed to deploy the contract, which is essentially the
+            # byte code parameter in the compiled asset JSON
+            f.write(f'  auto data_{k}_deploy() -> cbdc::buffer {{\n')
+            f.write(f'    auto buf = cbdc::buffer::from_hex("{v["bytecode"][2:]}");\n')
+            f.write('    return buf.value();\n')
+            f.write('  }\n\n')
+
+            # Loop over the functions in the ABI
+            for abi in v['abi']:
+                # Only make methods for functions, ignore events (for now)
+                if abi['type'] == 'function':
+                    # Write the method name data_<shortname>_<methodname>
+                    f.write(f'auto data_{k}_{camel_to_snake(abi["name"])}(')
+                    # Write all parameters as function arguments
+                    for idx, inp in enumerate(abi['inputs']):
+                        tp = 'bytes32'
+                        if inp['type'] == 'uint256':
+                            tp = 'uint256be'
                         if inp['type'] == 'address':
-                            f.write('  buf.extend(address_param_offset);\n')
-                        f.write('  buf.append({name}.bytes, sizeof({name}.bytes));\n'.format_map(dict({'name':to_snake(inp['name'])})))
-
-                # Return the buffer we built
-                f.write('    return buf;\n')
-                f.write('  }\n\n')
-
-    f.write('}\n\n')
-    f.write('#endif // OPENCBDC_TX_TOOLS_BENCH_PARSEC_EVM_CONTRACTS_H_\n\n')
-
+                            tp = 'address'
+                        if idx > 0:
+                            f.write(', ')
+                        f.write(f'evmc::{tp} {camel_to_snake(inp["name"])}')
+
+                    # Write the return method and creation of the empty buffer
+                    f.write(') -> cbdc::buffer {\n')
+                    f.write('    auto buf = cbdc::buffer();\n')
+
+                    # Write the method selector calculation
+                    f.write('    const auto selector_{name} = std::string("{name_raw}('.format_map(dict({'name':camel_to_snake(abi['name']),'name_raw':abi['name']})))
+                    for idx, inp in abi['inputs']:
```

Comment:
This should be `enumerate(abi['inputs'])` if you would like to iterate through this this way